dashboard/plugins/s3_usage/s3_stats.py

```diff
@@ -0,0 +1,135 @@
+import boto3
+import sqlite3
+import fcntl
+import os
+import time
+import threading
+import re
+from collections import defaultdict
+
+class S3Stats:
+    def __init__(self, logger, params):
+        self.logger = logger
+        self.params = params
+        self.validate_params()
+
+    def validate_params(self):
+        if not 'ttl' in self.params:
+            self.params['ttl'] = 60 * 60 * 24 # 1 day
+        if not 'check_every' in self.params:
+            self.params['check_every'] = 60 * 60 # 1 hour
+
+    def is_ready(self):
+        return os.path.exists(S3StatsRefreshTask.STATS_DB_PATH)
+
+    def query(self, base, depth=None):
+        if depth is None:
+            depth = len(base.split('/')) - 1
+        sqlite = sqlite3.connect(S3StatsRefreshTask.STATS_DB_PATH)
+        sqlite.row_factory = sqlite3.Row
+        cursor = sqlite.cursor()
+        cursor.execute(f"SELECT * FROM stats WHERE path like '{base}%' and depth = {depth} order by size desc")
+        result = cursor.fetchall() 
+        cursor.close()
+        sqlite.close()
+        return result
+
+    def run_refresher_task(self):
+        S3StatsRefreshTask(self.logger, **(self.params)).run()
+        threading.Timer(self.params['check_every'], self.run_refresher_task).start()
+
+    def start_refresher(self):
+        threading.Timer(1, self.run_refresher_task).start() 
+
+class S3StatsRefreshTask:
+    STATS_DB_PATH = '/var/run/s3_stats.db'
+    STATS_DB_TMP_PATH = '/var/run/s3_stats.db.tmp'
+    LOCK_PATH = '/var/run/s3_stats.db.lock'
+    CREATE_STMT = 'CREATE TABLE stats (path VARCHAR, files INTEGER, size INTEGER, depth INTEGER)'
+    INSERT_STMT = 'INSERT INTO stats VALUES (?, ?, ?, ?)'
+
+    def __init__(self, logger, buckets_regexp, aws_access_key_id, aws_secret_access_key, check_every, ttl):
+        s3 = boto3.client('s3', 
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key)
+        self.s3_aggregator = S3Aggregator(s3)
+        self.buckets = [bucket['Name'] for bucket in s3.list_buckets()['Buckets'] 
+                                            if re.match(buckets_regexp, bucket['Name'])]
+        self.logger = logger
+        self.ttl = ttl
+
+    def needs_refresh(self):
+        try:
+            return time.time() - os.stat(self.STATS_DB_PATH).st_mtime > self.ttl
+        except FileNotFoundError:
+            return True
+
+    def run(self):
+        with open(self.LOCK_PATH, 'w') as lock:
+            self.logger.debug("Trying to acquire the S3Stats lock")
+            fcntl.flock(lock, fcntl.LOCK_EX)
+            self.logger.debug("S3Stats Lock acquired")
```

Comment:
I'm assuming this lock also blocks reads for the time of writing? Not sure how long writing takes, but if you want to be super fancy and avoid locking, then maybe instead you could simply create a new temp file db and then when writing is finished make a rename to swap the original stale db. This way next query will use the new file and the old one can be removed. Just an idea :)

Reply:
This is exactly how the refresh mechanism is implemented :-) One worker takes the lock, rebuilds database on path `/var/run/s3_stats.db.tmp` and when it's finished, replaces `/var/run/s3_stats.db`. From frontend perspective - the database is opened on each query, so when the swap happens, we don't need to refresh anything in `views.py`, because the next request will go to newer file.

Reply:
Ups I missed the tmp/rename somehow, good to hear we are like minded :D

Reply:
👥  ;-)